sweep/sweeper.go

```diff
@@ -259,23 +276,40 @@ func (s *UtxoSweeper) Stop() error {
 }
 
 // SweepInput sweeps inputs back into the wallet. The inputs will be batched and
-// swept after the batch time window ends.
+// swept after the batch time window ends. A custom fee preference can be
+// provided, otherwise the UtxoSweeper's default will be used.
 //
 // NOTE: Extreme care needs to be taken that input isn't changed externally.
 // Because it is an interface and we don't know what is exactly behind it, we
 // cannot make a local copy in sweeper.
-func (s *UtxoSweeper) SweepInput(input input.Input) (chan Result, error) {
+func (s *UtxoSweeper) SweepInput(input input.Input,
+	feePreference *FeePreference) (chan Result, error) {
+
 	if input == nil || input.OutPoint() == nil || input.SignDesc() == nil {
 		return nil, errors.New("nil input received")
 	}
 
+	// If the client didn't specify a fee preference, we'll use the
+	// sweeper's default. Otherwise, we'll ensure they've provided a valid
+	// fee preference.
+	if feePreference == nil {
+		feePreference = &FeePreference{
+			ConfTarget: s.cfg.SweepTxConfTarget,
+		}
+	}
+	feeRate, err := DetermineFeePerKw(s.cfg.FeeEstimator, *feePreference)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Could the whole idea of conf targets be moved outside of `UtxoSweeper`?

Reply:
As in allow the sweeper to only take in fee rates, so the caller would be responsible for mapping a target to a fee rate?

Reply:
IMO it make sense to have it live here. Callers are expressing their time preference w.r.t their expectation of when the sweep would happen. Pushing this mapping on the callers would just add code duplication at each call site, when it can instead all be consolidated here.

Reply:
Yeah it's necessary to have it done at the `UtxoSweeper` level anyway in order to get the confirmation target.

Reply:
I was thinking about a wrapper struct around the `UtxoSweeper` for callers that want to specify a conf target. So callers still don't need to do the mapping themselves.

`UtxoSweeper` isn't the easiest code to understand, so splitting off what's possible may help.

But not a big issue.

Reply:
Yeah agree that it would be a nice separation of concerns. Seems like a good candidate for a follow-up PR though?

Reply:
I think we have to be realistic at follow up, pure refactor prs. They are mostly not going to happen. Also we would incur a risk twice. First changes in this pr and when that is all good, we change the system again with new risks.

Imo it is better to try to do it first time right. Less overall costs. We surely don't want to over-engineer, but the size of the sweeper as it currently is justifies some effort in separation responsibilities. Costs over the life time of the software of future devs trying to understand code adds up.

An alternative design that could be good is to offer inputs to `UtxoSweeper` with a `GetFeeRate` callback. That way callers can pass in a function that either returns a static fee rate, a fee rate calculated from a conf target or something more advanced that also takes into account the value of the input for example. For `UtxoSweeper` it'll be all the same. When it sweeps, it just collects fee rates and does its clustering magic. 

For conf targets, there could be an object `feeEstimationCache` that caches the fee rate estimates for the block. The `getFeeRate()` callback for an input could be something like this closure:
`func() {return feeEstimationCache.GetFeeRate(defaultCommitSweepTarget)}`

Reply:
IMO we shouldn't push all this burden on the caller. The goal is to be able to allow the caller to express a fee preference. In the end, that preference may not be _exactly_ met, but within a ball park range, particularly if there's a slightly different mapping that results in `lnd` paying lower fees over all. The code in question as is is also just a few lines: calling to the estimator. If we want to add some more magic between the caller and the estimator, then we can do that at the estimator interface level. 

If we look at the current use cases within `lnd`, every time a caller goes to request that something be swept, it has a rough _time frame_ w.r.t when it should be swept. Scenarios such as sweeping an HTLC have a very clear time frame (to avoid the race), while presences such as sweeping a CSV output on a confirmed output don't (as long as it isn't a revoked commitment ;)). 

> Yeah agree that it would be a nice separation of concerns. 
> but the size of the sweeper as it currently is justifies some effort in separation responsibilities.

The prior approach had an even more direct separation of concerns, but here we are now. Let's not forget the reality of things as they exist today: chain fess are rising, and users need to be able to specify a time preference. Even our own systems need this capability in order to scale and be sustainable. We have an opportunity to get something incrementally better in (atm we basically have nothing) before the next scheduled major release. Lets focus on the use cases at hand that we've experienced the need for first had (and has also been requested time after time by users of `lnd`). 

> An alternative design that could be good is to offer inputs to UtxoSweeper with a GetFeeRate callback

In the near future, do we ever see this being anything other than the default? All our current use cases just want to express a preference, and have the fee rate be _loosely_ adjusted (there're no guarantees after all) if that is at risk of not being achieved. Down the line if we have a compelling use case, this can be added in: for these special inputs, the sweeper skips its regular mapping and uses that function instead.

Reply:
>The code in question as is is also just a few lines: calling to the estimator. 

It is more than a few lines, it is also the re-targeting part. I think an object shouldn't have too many responsibilities and the conf target mapping is something that can be extracted. But in the end it comes down to a believe about what is the best design.

>In the near future, do we ever see this being anything other than the default? 

As I now understand it, there are two implementations of this callback: either conf target based (returns a diff. fee rate every block) and a static fee rate (passed in on the cmd line).

Reply:
I believe this could be both simplified and made more safe for many of our use cases by specifying an _absolute block height conf target_ instead the current conf target in number of blocks. Say we need a sweep to confirm by a specific height, the sweeper would start out low and increase the feerate as the target approaches.

Most important IMO is to adjust the fee rate dynamically, which could be one of the really powerful usecases of a sweeper like this.